### PR TITLE
Modular block request handler

### PR DESCRIPTION
**Pull request by rahulksnv**
If a substrate user wants to implement a custom block relay protocol (like Compact Blocks, Graphene, etc), it is not currently possible. This is because the block request handler is tightly coupled with the rest of the stack. This change makes the block request handler modular.

Main changes:
1. `ChainSync` currently calls into the block request handler directly. Instead, move the block request handler behind a trait. This allows new protocols to be plugged into ChainSync.
2. `BuildNetworkParams` is changed so that custom relay protocol implementations can be (optionally) passed in during network creation time. If custom protocol is not specified, it defaults to the existing block handler
3. `BlockServer` and `BlockDownloader` traits are introduced for the protocol implementation. The existing block handler has been changed to implement these traits
4. Other changes: 
   - make `TxHash` serializable. This is needed for exchanging the serialized hash in the relay protocol messages
   - clean up types no longer used(`OpaqueBlockRequest`, `OpaqueBlockResponse`)
   
 polkadot companion:  https://github.com/paritytech/polkadot/pull/7134
 cumulus companion: https://github.com/paritytech/cumulus/pull/2489

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2730016

File: client/network/sync/src/lib.rs
Comment (dmitry-markin) [inline client/network/sync/src/lib.rs:2132]: Is it intentional that we don't disconnect a peer in case of failed extraction, like it was with `decode_block_response`?

**Review by dmitry-markin — APPROVED** [review]
Looks good, just one minor comment. Thanks!

File: client/network/sync/src/block_relay_protocol.rs
Comment (altonen) [inline client/network/sync/src/block_relay_protocol.rs:32]: There is a problem with the current block request handler which is that even if `SyncingEngine` has rejected the peer, it's still able to download blocks because the request handler doesn't know which peers have been accepted and which have been rejected. As long as you have a connection open to the peer, you can download blocks from them.

It's not necessarily an issue with this PR but this bug has to be fixed meaning I wouldn't necessarily build on the current design of the block request handler because it has to be rewritten and coupled more tightly with `SyncingEngine`.

File: client/network/sync/src/block_relay_protocol.rs
Comment (altonen) [inline client/network/sync/src/block_relay_protocol.rs:42]:
```suggestion
	/// can be extracted from this response using `BlockDownloader::block_response_into_blocks()`.
```

File: client/network/sync/src/block_relay_protocol.rs
Comment (altonen) [inline client/network/sync/src/block_relay_protocol.rs:59]:
```suggestion
pub enum BlockResponseError {
```

File: client/network/sync/src/block_request_handler.rs
Comment (altonen) [inline client/network/sync/src/block_request_handler.rs:515]: This whole code could be cleaned up with `then_some()`

File: client/network/sync/src/lib.rs
Comment (altonen) [inline client/network/sync/src/lib.rs:1295]: Why is a handle to `NetworkServiceProvider` not stored inside the `BlockDownloader`? Making a copy of each for every request doesn't seem right.

Making a copy of either one of them seems weird. Why can't `BlockDownloader` be some opaque service `ChainSync` calls in order to downloads blocks?

Comment (rahulksnv) [reply client/network/sync/src/lib.rs:1295]: Yes, changed the `BlockDownloader` to take the network handle during init time

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2737684

File: client/network/sync/src/lib.rs
Comment (rahulksnv) [reply client/network/sync/src/lib.rs:2131]: Yes, the change tries to maintain the existing behavior

Before the change:
1. On decode failure: https://github.com/paritytech/substrate/blob/master/client/network/sync/src/lib.rs#L2208, it both disconnects and reports the peer
2. On extraction failure: https://github.com/paritytech/substrate/blob/943c520aa78fcfaf3509790009ad062e8d4c6990/client/network/sync/src/lib.rs#L2075, it only reports the peer, but does not disconnect

`BlockResponseError` was added specifically to distinguish between these two cases

File: client/network/sync/src/block_relay_protocol.rs
Comment (rahulksnv) [reply client/network/sync/src/block_relay_protocol.rs:31]: >It's not necessarily an issue with this PR but this bug has to be fixed meaning I wouldn't necessarily build on the current design of the block request handler because it has to be rewritten and coupled more tightly with SyncingEngine.

Got it. Sounds like the rewrite will take a while. Since we are blocked on this change, can we get this in to make progress for now?

File: client/network/sync/src/block_request_handler.rs
Comment (rahulksnv) [reply client/network/sync/src/block_request_handler.rs:517]: I made the change which caused tests to fail. The eager evaluation of `then_some()` turned out to be the issue(took some binary search to figure it out). The APIs like `decode_all()` get called even when they should not be (e.g) empty `vec` and return error, etc causing the block extraction to fail. Using `then()` is not so ergonomic in this context, so I rolled back the change.

**Comment by rahulksnv** [comment]
Please point me how to fix the failing cumulus test (could not find anything in the contributing doc about making a dependent cumulus PR)

**Comment by altonen** [comment]
> Please point me how to fix the failing cumulus test (could not find anything in the contributing doc about making a dependent cumulus PR)

You need to make a cumulus companion, much the same way as you made the polkadot companion

File: client/network/sync/src/block_relay_protocol.rs
Comment (altonen) [reply client/network/sync/src/block_relay_protocol.rs:31]: Blocked how? As in you need this in Substrate ASAP?

File: client/network/sync/src/lib.rs
Comment (altonen) [reply client/network/sync/src/lib.rs:1294]: I'm not entirely sure why the code has been written this way but `SyncingEngine` shouldn't be sending requests to peers it doesn't know about. I think what previously happened if the peer was not known, the request was dropped as the RX half is dropped. I don't imagine this will cause issues

File: client/network/sync/src/block_relay_protocol.rs
Comment (rahulksnv) [reply client/network/sync/src/block_relay_protocol.rs:31]: Blocked as in: our stack is expected to be used in bandwidth constrained environments. The duplicate gossip (once as the transactions, again as the block extrinsics) make some (gossip hungry) features in our roadmap infeasible. The savings from the dedupe would go a long way. Hope that answers your question

Comment (altonen) [reply client/network/sync/src/block_relay_protocol.rs:31]: I don't have any qualms about the implementation (apart from the ones that already exist there) but I would rather merge this only after we've released v0.9.42 which I believe happens soon.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2800902

**Comment by altonen** [comment]
You need to merge master to polkadot/cumulus companions

**Review by melekes — APPROVED** [review]
lgtm 👍

**Comment by rahulksnv** [comment]
Hi @altonen @melekes @dmitry-markin Can you please let me know the next steps here, in terms of getting this merged? Sorry I am not clear about the process, as I am submitting an upstream patch for the first time. Thanks!

**Review by dmitry-markin — APPROVED** [review]

**Comment by dmitry-markin** [comment]
> Hi @altonen @melekes @dmitry-markin Can you please let me know the next steps here, in terms of getting this merged? Sorry I am not clear about the process, as I am submitting an upstream patch for the first time. Thanks!

Hi @rahulksnv, please merge master into you PR to resolve the merge conflicts. Than I would wait for @altonen approval, and we can merge the PR.

**Review by altonen — APPROVED** [review]

**Comment by rahulksnv** [comment]
> > Hi @altonen @melekes @dmitry-markin Can you please let me know the next steps here, in terms of getting this merged? Sorry I am not clear about the process, as I am submitting an upstream patch for the first time. Thanks!
> 
> Hi @rahulksnv, please merge master into you PR to resolve the merge conflicts. Than I would wait for @altonen approval, and we can merge the PR.

 @dmitry-markin merged with latest master, thanks

**Comment by rahulksnv** [comment]
> > @dmitry-markin merged with latest master, thanks
> 
> Could you do the same with companion PRs for tests to pass?

Re-synced all three repos the corresponding master.

**Comment by zk-qed** [comment]
> Hey, is anyone still working on this? Due to the inactivity this issue has been automatically marked as stale. It will be closed if no further activity occurs. Thank you for your contributions.

Thanks for the reminder. I have been on a break, will wrap this up (resync with latest master)

**Comment by rahulksnv** [comment]
> Hey, is anyone still working on this? Due to the inactivity this issue has been automatically marked as stale. It will be closed if no further activity occurs. Thank you for your contributions.

Thanks for the reminder, was on a break, will resync with master and try to wrap this up

**Pull request closed without merging**
